src/community/jdbcstore/src/main/java/org/geoserver/jdbcstore/internal/JDBCQueryHelper.java

```diff
@@ -0,0 +1,568 @@
+            s.setBinaryStream(index, value);
+        }
+
+        @Override
+        public InputStream getValue(ResultSet rs, String colName) throws SQLException {
+            return rs.getBinaryStream(colName);
+        }        
+    };
+    
+    /**
+     * 
+     * Query Field.
+     *
+     * @param <T>
+     */
+    public static class Field<T> {
+        private final String fieldName;
+        private final String fieldExpression;
+        private final Type<T> type;
+                
+        public Field(String fieldName, String fieldExpression, Type<T> type) {
+            super();
+            this.fieldName = fieldName;
+            this.fieldExpression = fieldExpression;
+            this.type = type;
+        }
+        public String getFieldName() {
+            return fieldName;
+        }
+        public String getFieldExpression() {
+            return fieldExpression;
+        }
+        public Type<T> getType() {
+            return type;
+        }
+        
+        public T getValue(ResultSet rs) throws SQLException {
+            return type.getValue(rs, fieldName);
+        }
+    }
+    
+    /**
+     * 
+     * Query parameter.
+     *
+     * @param <T>
+     */
+    public static class Parameter<T> {
+        private final Type<T> type;
+        private final T value;        
+        public Parameter(Type<T> type, T value) {
+            this.type = type;
+            this.value = value;
+        }
+        public Type<T> getType() {
+            return type;
+        }
+        public T getValue() {
+            return value;
+        }                
+        public void setValue(PreparedStatement s, int index) throws SQLException {
+            type.setValue(s, index, value);
+        }
+    }
+    
+    /**
+     * Builds a prepared statement from a query
+     *
+     */
+    public static class QueryBuilder {
+        private StringBuilder stringBuilder = new StringBuilder();
+        private ArrayList<Parameter<?>> parameters = new ArrayList<Parameter<?>>();
+        
+        public QueryBuilder() {}
+        
+        public QueryBuilder(String string) {
+            stringBuilder.append(string);
+        }
+
+        public void append(String s) {
+            stringBuilder.append(s);
+        }
+        
+        public void addParameter(Parameter<?> parameter) {
+            parameters.add(parameter);
+        }
+        
+        private void parameters(PreparedStatement ps) throws SQLException {
+            for (int i = 0; i < parameters.size(); i++) {
+                parameters.get(i).setValue(ps, i + 1);
+            }
+        }
+        
+        public PreparedStatement toStatement(Connection c) throws SQLException {
+            PreparedStatement ps = c.prepareStatement(stringBuilder.toString());
+            try {
+                parameters(ps);
+                return ps;
+            } catch (SQLException e) {
+                ps.close();         
```

Comment:
It shouldn't. The PreparedStatement is being returned, so it should not be closed.